query/expression.go

```diff
@@ -144,19 +144,66 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
 }
 
 func (expr *functionExpression) Evaluate(context EvaluationContext) (value, error) {
-	switch expr.functionName {
+
+	name := expr.functionName
+	switch name {
 	case "+":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left + right })
+		fallthrough
 	case "-":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left - right })
-	case "/":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left / right })
+		fallthrough
 	case "*":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left * right })
+		fallthrough
+	case "/":
+		arguments := make([]value, len(expr.arguments))
+		var err error
+		for i := range arguments {
+			arguments[i], err = expr.arguments[i].Evaluate(context)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Evaluation of a binary operator:
+		if len(arguments) != 2 {
+			return nil, errors.New(fmt.Sprintf("Function `%s` expects 2 operands but received %d (%+v)", name, len(arguments), arguments))
+		}
+		left := arguments[0]
+		right := arguments[1]
+		operatorMap := map[string]func(float64, float64) float64{
+			"+": func(x, y float64) float64 { return x + y },
+			"-": func(x, y float64) float64 { return x - y },
+			"*": func(x, y float64) float64 { return x * y },
+			"/": func(x, y float64) float64 { return x / y },
+		}
+		return evaluateBinaryOperation(context, name, left, right, operatorMap[name])
+	case "aggregate.sum":
+		fallthrough
+	case "aggregate.mean":
+		fallthrough
+	case "aggregate.min":
+		fallthrough
+	case "aggregate.max":
+		arguments := make([]value, len(expr.arguments))
+		var err error
+		for i := range arguments {
+			arguments[i], err = expr.arguments[i].Evaluate(context)
+			if err != nil {
+				return nil, err
+			}
+		}
+		funMap := map[string]aggregate{
+			"aggregate.sum":  sumAggregate,
+			"aggregate.mean": meanAggregate,
+			"aggregate.min":  minAggregate,
+			"aggregate.max":  maxAggregate,
+		}
+		if len(arguments) != 1 {
+			return nil, errors.New(fmt.Sprintf("Function `%s` expected 2 operands but received %d (%+v)", name, len(arguments), arguments))
+		}
+		list, err := arguments[0].toSeriesList(context.Timerange)
+		if err != nil {
+			return nil, err
+		}
+		return seriesListValue(aggregateBy(list, aggregateMap[funMap[name]], expr.groupBy)), nil
```

Comment:
`aggregateMap` here is really confusing since it appears to come out of nowhere. The `*Aggregate` enums also do the same... Maybe we should move these function implementations to their own package or something? How things are set up makes some of these references a little too magic for my liking, though maybe that's just how go works.

@jeeyoungk